### PR TITLE
Made CryptoException extend IOException

### PR DESCRIPTION
**Pull request by dlmarion**
Modified CryptoException such that it no longer extends RuntimeException and extends IOException instead. Initially I was going to modify it to extend GeneralSecurityException but that touched a lot of code and caused issues with Accumulo classes that implemented other interfaces that declared only IOException.

### CODE REVIEW

**Comment by dlmarion** [comment]
Merging this issue should complete the subtask "Make CryptoException extend GeneralSecurityException" in #2930

File: core/src/main/java/org/apache/accumulo/core/spi/crypto/PerTableCryptoServiceFactory.java [hunk 1]
```diff
@@ -46,12 +48,24 @@ public class PerTableCryptoServiceFactory implements CryptoServiceFactory {
...
-  public CryptoService getService(CryptoEnvironment environment, Map<String,String> props) {
+  public CryptoService getService(CryptoEnvironment environment, Map<String,String> props)
+      throws CryptoException {
     if (environment.getScope() == CryptoEnvironment.Scope.WAL) {
-      return cryptoServiceMap.computeIfAbsent(WAL_FAKE_ID, (id) -> getWALServiceInitialized(props));
+      return cryptoServiceMap.computeIfAbsent(WAL_FAKE_ID, (id) -> {
+        try {
+          return getWALServiceInitialized(props);
+        } catch (CryptoException e) {
+          throw new UncheckedIOException("Error creating WAL crypto service", e);
```
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/spi/crypto/PerTableCryptoServiceFactory.java:58]: If getService() is already throwing CryptoException, why catch and rethrow as Unchecked?

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/spi/crypto/PerTableCryptoServiceFactory.java:58]: It's for the lambda, can't throw checked exceptions

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/crypto/PerTableCryptoServiceFactory.java:58]: OK that makes sense. Check out my other comment about checked vs unchecked.

**Comment by dlmarion** [comment]
> [This](https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html) is a nice clarification from Oracle docs: "If a client can reasonably be expected to recover from an exception, make it a checked exception. If a client cannot do anything to recover from the exception, make it an unchecked exception."

Do you think this should not be merged then and leave CryptoException as a RuntimeException. I was thinking that having a known checked Exception in the SPI methods makes sense so that errors specific to crypto can be handled.

**Comment by milleruntime** [comment]
> Do you think this should not be merged then and leave CryptoException as a RuntimeException. I was thinking that having a known checked Exception in the SPI methods makes sense so that errors specific to crypto can be handled.

For most situations, I don't think there will be much we can do if we catch a CryptoException (coming from a TabletServer, for example) other than restart the server. Since these interfaces are in the SPI, the "client" who handles the exception will usually be server code. And with the way the CryptoService runs within the tserver, there is a good chance a restart will have to happen to reload the service. Eventually, once the Crypto is properly tested we could drop the experimental tag and be able to determine a better way of handling the errors. Or if we create a designated Server running just for Crypto, there may be better ways to handle the errors. The RFile API may be a good place to throw a checked exception for example, allowing the client to handle encryption errors.

I do think there is room for improving the overall error handling though. But maybe changing the exception type isn't the right approach.

**Comment by dlmarion** [comment]
We are going to leave CryptoException as a RuntimeException per comment in #2930

**Pull request closed without merging**